dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusAArch64.java

```diff
@@ -105,163 +143,302 @@ public Object visitInstructionRow(LitmusAArch64Parser.InstructionRowContext ctx)
-        return programBuilder.addChild(mainThread, EventFactory.newLocal(register, expr));
+        return add(EventFactory.newLocal(register, expr));
     }
 
     @Override
-    public Object visitCmp(LitmusAArch64Parser.CmpContext ctx) {
+    public Object visitCmp(CmpContext ctx) {
         Register register = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
         Expression expr = ctx.expr32() != null ? (Expression)ctx.expr32().accept(this) : (Expression)ctx.expr64().accept(this);
         lastCmpInstructionPerThread.put(mainThread, new CmpInstruction(register, expr));
         return null;
     }
 
     @Override
-    public Object visitArithmetic(LitmusAArch64Parser.ArithmeticContext ctx) {
-        Register rD = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
-        Register r1 = programBuilder.getOrErrorRegister(mainThread, ctx.rV);
-        Expression expr = ctx.expr32() != null ? (Expression)ctx.expr32().accept(this) : (Expression)ctx.expr64().accept(this);
-        return programBuilder.addChild(mainThread, EventFactory.newLocal(rD, expressions.makeIntBinary(r1, ctx.arithmeticInstruction().op, expr)));
+    public Object visitArithmetic(ArithmeticContext ctx) {
+        final Register r64 = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
+        final Register register = getOrNewRegister32(ctx.rD32, r64, false, false);
+        final Register operand = programBuilder.getOrErrorRegister(mainThread, ctx.rV);
+        final Expression expr = (Expression) (ctx.expr32() != null ? ctx.expr32() : ctx.expr64()).accept(this);
+        final Expression fittedOperand = expressions.makeCast(operand, expr.getType());
+        final Expression result = expressions.makeIntBinary(fittedOperand, ctx.arithmeticInstruction().op, expr);
+        add(EventFactory.newLocal(register, expressions.makeCast(result, register.getType())));
+        addRegister64Update(r64, register);
+        return null;
     }
 
     @Override
-    public Object visitLoad(LitmusAArch64Parser.LoadContext ctx) {
-        Register register = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
-        Register address = programBuilder.getOrErrorRegister(mainThread, ctx.address().id);
-        if(ctx.offset() != null){
-            address = visitOffset(ctx.offset(), address);
-        }
-        return programBuilder.addChild(mainThread, EventFactory.newLoadWithMo(register, address, ctx.loadInstruction().mo));
+    public Object visitLoad(LoadContext ctx) {
+        final Register r64 = parseRegister64(ctx.rD32, ctx.rD64);
+        final LoadInstructionContext inst = ctx.loadInstruction();
+        final Register register = getOrNewRegister32(ctx.rD32, r64, inst.byteSize, inst.halfWordSize);
+        final Expression address = parseAddress(ctx.address());
+        final String mo = inst.acquire ? MO_ACQ : MO_RX;
```

Comment:
This will treat plain loads as atomic relaxed. Can we always do this in arm? (it might be the case since there is no notion of data races thus plain and relaxed might be pretty much the same from the point of view of the WMM)

Reply:
I agree that this is incorrect. We shouldn't generate tagging that is not used by the WMM.